### PR TITLE
[Fix] Fix num_last_epochs logic for the hooks in YOLOX training

### PR DESCRIPTION
**Pull request by HJoonKwon**
## Motivation

Based on #7284 , I ran a few more experiments with ``` num_last_epochs```. 
I could still see that the ``` before_train_epoch``` hooks didn't come in at the proper epoch. 
They came in one epoch earlier than I expected them to, so I made a PR to fix it. 

On the other hand, the learning rate schedule, which is the learning rate is fixed during the ``` num_last_epochs```, works fine. 
This means that the ``` number of epochs``` for fixed learning rate and the ``` number of epochs``` for non-mixup-mosaic augmentation didn't match as well. 

## Modification

I modified two files, ``` sync_norm_hook.py``` and  ``` yolox_mode_switch_hook.py``` that implement ``` before_train_epoch``` hook using ``` num_last_epochs```. 
In both of the files, the if logic was modified from
```python
        if (epoch + 1) == runner.max_epochs - self.num_last_epochs:
```
to
```python
        if epoch == runner.max_epochs - self.num_last_epochs:
```

## BC-breaking (Optional)

Yes

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by shinya7y** [comment]
The official YOLOX code also adds one more epoch. https://github.com/Megvii-BaseDetection/YOLOX/pull/1152 proposes a fix for the code.
Since the authors have not commented so far, we cannot determine whether it is a bug or an intentional specification.

If it is intentional,
`num_last_epochs=15` -> `num_last_epochs=16` in `yolox_mode_switch_hook.py`, `sync_norm_hook.py`

**Comment by shinya7y** [comment]
This PR makes a BC-breaking change.
If we want to reproduce the settings of the published trained models, we need to modify `yolox_s_8x8_300e_coco.py`.
For users that use their own configs, it is worth considering raising warnings.

**Comment by HJoonKwon** [comment]
@shinya7y The author in the [paper](https://arxiv.org/abs/2107.08430) says 

> We adopt the MixUp and Mosaic implementation in our model and close it for the last 15 epochs

So I think it's a bug but let's wait for the answer. 

What a coincidence that person made a PR 1 day after I published an issue here :)

**Comment by hhaAndroid** [comment]
@HJoonKwon Sorry for the late reply. Although the understanding is a bit wrong, in fact, this is not a bug, this is completely according to the official way of writing. If num_last_epochs=15, it means switching from 300-15

**Comment by HJoonKwon** [comment]
@hhaAndroid Still confused.. Switching from 300-15(in the meta data perspective) means that training closes the augmentation from 285th epoch to 300th epoch. And that is 16 epochs, not 15 epochs. There might be some confusion in the '[writing](https://github.com/open-mmlab/mmdetection/pull/7322#issuecomment-1059692767)' in the paper. But, sure we may wait! :)

**Comment by nemonameless** [comment]
> 

Have you tried it? Are there any differences in the results？

**Comment by voldemortX** [comment]
While the official code still haven't updated. I would suggest that the `+1` be removed from `num_last_epochs` , and setting it to `16` in the configs with a comment that leads to these discussions.

This could be extra confusing as mentioned in #8560, when people expect to tune off this thing through `num_last_epochs=0`.
